### PR TITLE
AMB Upgrade + BAL Pipes replace

### PR DESCRIPTION
**Pull request by viacheslavvbogdanov**
see AMBUpgrade_BalPipesReplaceCheck.ts

### CODE REVIEW

File: contracts/base/strategies/multi/pipelines/LinearPipeline.sol [hunk 1]
```diff
@@ -128,29 +129,77 @@ contract LinearPipeline is ILinearPipeline, Initializable {
...
       newPipe.setPrevPipe(address(prevPipe));
     } else {// first pipe should have pipeline as prev pipe to send tokens on get()
       newPipe.setPrevPipe(address(this));
     }
   }
 
+  /// @dev Replaces pipe in the pipeline
+  /// @param pipeIndex index of pipe to change
+  /// @param newPipe to be changed
+  /// @param maxDecrease1000 maximum decrease ot totalAmountOut (in 0,1%)
```
Comment (belbix) [inline contracts/base/strategies/multi/pipelines/LinearPipeline.sol:150]: type ot => of

File: contracts/strategies/matic/multi/StrategyAaveMaiBal.sol [hunk 2]
```diff
@@ -17,7 +17,7 @@ import "../../../base/SlotsLib.sol";
 /// @title AAVE->MAI->BAL Multi Strategy Polygon Implementation
 /// @author belbix, bogdoslav
 contract StrategyAaveMaiBal is AaveMaiBalStrategyBase {
-  using SlotsLib for bytes32;
+  using SlotsLib for uint;
```
Comment (belbix) [inline contracts/strategies/matic/multi/StrategyAaveMaiBal.sol:20]: you can remove it here

File: contracts/strategies/matic/multi/StrategyMaiBal.sol [hunk 3]
```diff
@@ -17,7 +17,7 @@ import "../../../base/SlotsLib.sol";
 /// @title MAI->BAL Multi Strategy Polygon Implementation
 /// @author belbix, bogdoslav
 contract StrategyMaiBal is MaiBalStrategyBase {
-  using SlotsLib for bytes32;
+  using SlotsLib for uint;
```
Comment (belbix) [inline contracts/strategies/matic/multi/StrategyMaiBal.sol:20]: remove

File: contracts/strategies/matic/multi/StrategyAaveMaiBal.sol (hunk 2 above)
Comment (viacheslavvbogdanov) [reply contracts/strategies/matic/multi/StrategyAaveMaiBal.sol]: Remove what? SlotsLib I need to set _MAI_CAM_PIPE_SLOT  & _MAI_STABLECOIN_PIPE_SLOT

File: contracts/strategies/matic/multi/StrategyMaiBal.sol (hunk 3 above)
Comment (viacheslavvbogdanov) [reply contracts/strategies/matic/multi/StrategyMaiBal.sol]: It needed for _MAI_STABLECOIN_PIPE_SLOT

File: contracts/strategies/matic/multi/StrategyAaveMaiBal.sol (hunk 2 above)
Comment (viacheslavvbogdanov) [reply contracts/strategies/matic/multi/StrategyAaveMaiBal.sol]: I going to remove it

File: contracts/strategies/matic/multi/StrategyMaiBal.sol (hunk 3 above)
Comment (viacheslavvbogdanov) [reply contracts/strategies/matic/multi/StrategyMaiBal.sol]: I going to remove it and replace to internal functions for do not update it when pipes are replaced

File: contracts/base/strategies/multi/pipelines/LinearPipeline.sol [hunk 4]
```diff
@@ -286,7 +329,4 @@ contract LinearPipeline is ILinearPipeline, Initializable {
     return abi.decode(reason, (uint256));
   }
```
Comment (belbix) [inline contracts/base/strategies/multi/pipelines/LinearPipeline.sol:331]: keep gap

**Pull request merged**